### PR TITLE
feat: add transaction

### PR DESCRIPTION
**Pull request by thevaibhav-dixit**
Rebase to main after #35 gets merged

### CODE REVIEW

File: cala-ledger-core-types/src/primitives.rs [hunk 1]
```diff
@@ -6,6 +6,8 @@ crate::entity_id! { AccountId }
 crate::entity_id! { JournalId }
 crate::entity_id! { DataSourceId }
 crate::entity_id! { TxTemplateId }
+crate::entity_id! { TransactionId }
+crate::entity_id! { CorrelationId }
```
[the remark this answers is not in the document]
Comment (thevaibhav-dixit) [reply cala-ledger-core-types/src/primitives.rs:10]: at least in the sqlx-ledger it is a separate type

File: cala-ledger/migrations/20231208110808_cala_ledger_setup.sql [hunk 2]
```diff
@@ -71,6 +71,29 @@ CREATE TABLE cala_tx_template_events (
   FOREIGN KEY (data_source_id, id) REFERENCES cala_tx_templates(data_source_id, id)
 );
 
+CREATE TABLE cala_transactions (
+  data_source_id UUID NOT NULL DEFAULT '00000000-0000-0000-0000-000000000000',
+  id UUID NOT NULL,
+  journal_id UUID NOT NULL,
```
[the remark this answers is not in the document]
Comment (thevaibhav-dixit) [reply cala-ledger/migrations/20231208110808_cala_ledger_setup.sql:77]: i think we need journal_id, transaction_id for filters and external_id and correlation_id for index

File: cala-ledger/src/transaction/mod.rs [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+            n_new_events,
+        } = self.repo.create_in_tx(&mut tx, new_transaction).await?;
+        self.outbox
+            .persist_events(tx, transaction.events.last_persisted(n_new_events))
+            .await?;
+        Ok(transaction)
+    }
+
+    #[instrument(name = "cala_ledger.transactions.list_by_template_id", skip(self))]
+    pub async fn list_by_template_id(
```
[the remark this answers is not in the document]
Comment (thevaibhav-dixit) [reply cala-ledger/src/transaction/mod.rs]: yeah you are right not required ATM. can bring it back when needed in future

File: cala-ledger/src/transaction/entity.rs [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+    pub(super) fn import(source: DataSourceId, values: TransactionValues) -> Self {
+        let events = EntityEvents::init(
+            values.id,
+            [TransactionEvent::Imported {
+                source: DataSource::Remote { id: source },
+                values,
+            }],
+        );
+        Self::try_from(events).expect("Failed to build transaction from events")
+    }
```
Comment (bodymindarts) [inline cala-ledger/src/transaction/entity.rs:50]: line break

File: cala-ledger/src/transaction/mod.rs [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+impl Transactions {
+    pub fn new(pool: &PgPool, outbox: Outbox) -> Self {
+        Self {
+            repo: TransactionRepo::new(pool),
+            outbox,
+            pool: pool.clone(),
+        }
+    }
```
Comment (bodymindarts) [inline cala-ledger/src/transaction/mod.rs:31]: line break

File: proto/ledger/outbox_service.proto [hunk 6]
```diff
@@ -116,3 +117,19 @@ enum ParamDataType {
...
+message TransactionCreated {
+  string data_source_id = 1;
+  Transaction transaction = 2;
+}
+
+message Transaction {
+  string id = 1;
+  string journal_id = 2;
+  string tx_template_id = 3;
+  string effective= 4;
```
Comment (bodymindarts) [inline proto/ledger/outbox_service.proto:130]: fmt

File: cala-ledger/migrations/20231208110808_cala_ledger_setup.sql [hunk 7]
```diff
@@ -71,6 +71,28 @@ CREATE TABLE cala_tx_template_events (
...
+  external_id VARCHAR DEFAULT NULL,
+  created_at TIMESTAMPTZ NOT NULL DEFAULT NOW(),
+  UNIQUE(data_source_id, id),
+  FOREIGN KEY (data_source_id, journal_id) REFERENCES cala_journals(data_source_id, id)
+);
+CREATE UNIQUE INDEX idx_cala_transactions_data_source_id_external_id ON cala_transactions (data_source_id, external_id) WHERE external_id IS NOT NULL;
```
Comment (bodymindarts) [inline cala-ledger/migrations/20231208110808_cala_ledger_setup.sql:83]: you also need one on `(data_source_id, id)`

File: cala-ledger/src/transaction/entity.rs [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+pub struct NewTransaction {
+    #[builder(setter(into))]
+    pub(super) id: TransactionId,
+    #[builder(setter(into))]
+    pub(super) journal_id: JournalId,
+    #[builder(setter(into))]
+    pub(super) tx_template_id: TxTemplateId,
+    pub(super) effective: chrono::NaiveDate,
+    #[builder(setter(into), default)]
+    pub(super) correlation_id: Option<String>,
```
Comment (bodymindarts) [inline cala-ledger/src/transaction/entity.rs:95]: I think we decided correlation_id is mandatory.

File: cala-ledger/src/transaction/mod.rs [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+    pub(crate) fn new(pool: &PgPool, outbox: Outbox) -> Self {
+        Self {
+            repo: TransactionRepo::new(pool),
+            outbox,
+            pool: pool.clone(),
+        }
+    }
+
+    #[instrument(name = "cala_ledger.transactions.create", skip(self))]
+    pub async fn create(
```
Comment (bodymindarts) [inline cala-ledger/src/transaction/mod.rs:34]: where is this being used? This should be illegal. You cannot create a transaction other than via `post_transaction` business logic... is this present in sqlx-ledger?

Comment (bodymindarts) [reply cala-ledger/src/transaction/mod.rs:34]: IF this is needed to be called from `post_transaction` then it must be `pub(crate)` simple `pub` is too broad...

**Pull request merged**